### PR TITLE
Keep track of resources.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: lib/ble.toit [hunk 1]
```diff
@@ -827,6 +984,12 @@ class Adapter:
 
   close -> none:
     if resource-group_:
```
Comment (kasperl) [inline lib/ble.toit:986]: Early exit?

File: lib/ble.toit [hunk 2]
```diff
@@ -393,10 +462,18 @@ class RemoteDevice extends Resource_:
 
   /**
   Disconnects from the remote device.
+
+  If $force is true, the connection is closed immediately. Otherwise, the
+    method waits until the remote device has acknowledged the disconnection.
   */
-  close:
+  close --force/bool=false:
+    services := discovered-services
```
Comment (kasperl) [inline lib/ble.toit:470]: It feels a bit icky that the field is public.

File: lib/ble.toit [hunk 3]
```diff
@@ -210,6 +220,19 @@ class RemoteCharacteristic extends RemoteReadWriteElement_ implements Attribute:
...
     super service characteristic
 
+  /**
+  Closes this characteristic.
+
+  Automatically removes the characteristic from the list of discovered characteristics of
+    the service.
+  */
+  close:
+    descriptors := discovered-descriptors
```
Comment (kasperl) [inline lib/ble.toit:230]: It feels a bit icky that the field is public. Maybe the field should be private and we should have a getter that returns a copy of the private list?

File: lib/ble.toit [hunk 4]
```diff
@@ -293,6 +319,12 @@ class RemoteCharacteristic extends RemoteReadWriteElement_ implements Attribute:
 
     return discovered-descriptors
 
+  /**
+  Removes the given remote-descriptor from the list of discovered descriptors.
```
Comment (kasperl) [inline lib/ble.toit:323]:
```suggestion
  Removes the given $remote-descriptor from the list of discovered descriptors.
```

File: lib/ble.toit [hunk 5]
```diff
@@ -343,14 +392,25 @@ class RemoteService extends Resource_ implements Attribute:
 
     return order-attributes_ characteristic-uuids discovered-characteristics
 
+  /**
+  Removes the given remote-characteristic from the list of discovered characteristics.
```
Comment (kasperl) [inline lib/ble.toit:396]:
```suggestion
  Removes the given $remote-characteristic from the list of discovered characteristics.
```

File: lib/ble.toit [hunk 6]
```diff
@@ -365,13 +425,22 @@ class RemoteDevice extends Resource_:
       throw "BLE connection failed"
       // TODO: Possible leak of the gatt_ resource on connection failures.
 
+  /**
+  Removes the given remote-service from the list of discovered services.
```
Comment (kasperl) [inline lib/ble.toit:429]:
```suggestion
  Removes the given $remote-service from the list of discovered services.
```

File: lib/ble.toit [hunk 7]
```diff
@@ -417,10 +494,28 @@ class LocalService extends Resource_ implements Attribute:
...
+  characteristics_/List := []
+
   constructor .peripheral-manager .uuid:
     resource := ble-add-service_ peripheral-manager.resource_ uuid.encode-for-platform_
     super peripheral-manager.adapter.resource-group_ resource --auto-release
 
+  /**
+  Closes this service and all its characteristics.
+  */
+  close:
```
Comment (kasperl) [inline lib/ble.toit:506]:
```suggestion
  close -> none:
```

File: lib/ble.toit [hunk 8]
```diff
@@ -417,10 +494,28 @@ class LocalService extends Resource_ implements Attribute:
...
+    characteristics := characteristics_
+    characteristics_ = []
+    characteristics.do: | characteristic/LocalCharacteristic | characteristic.close
+    peripheral-manager.remove-service_ this
+    close_
+
+  /**
+  Removes the given local-characteristic from the list of characteristics.
```
Comment (kasperl) [inline lib/ble.toit:514]:
```suggestion
  Removes the given $characteristic from the list of characteristics.
```

File: lib/ble.toit [hunk 9]
```diff
@@ -558,10 +655,28 @@ class LocalCharacteristic extends LocalReadWriteElement_ implements Attribute:
...
+  */
+  close:
+    descriptors := descriptors_
+    descriptors_ = []
+    descriptors.do: | descriptor/LocalDescriptor | descriptor.close
+    service.remove-characteristic_ this
+    close_
+
+  /**
+  Removes the given local-descriptor from the list of descriptors.
```
Comment (kasperl) [inline lib/ble.toit:675]:
```suggestion
  Removes the given $local-descriptor from the list of descriptors.
```

File: lib/ble.toit [hunk 10]
```diff
@@ -627,6 +742,13 @@ class LocalDescriptor extends LocalReadWriteElement_ implements Attribute:
     resource :=  ble-add-descriptor_ characteristic.resource_ uuid.encode-for-platform_ properties permissions value
     super characteristic.service resource
 
+  /**
+  Closes this descriptor.
+  */
+  close:
```
Comment (kasperl) [inline lib/ble.toit:748]:
```suggestion
  close -> none:
```

File: lib/ble.toit [hunk 11]
```diff
@@ -644,11 +766,19 @@ The manager for creating client connections.
...
   adapter/Adapter
 
+  remotes-devices_/List := []
+
   constructor .adapter:
     resource := ble-create-central-manager_ adapter.resource-group_
     super adapter.resource-group_ resource --auto-release
     resource-state_.wait-for-state STARTED-EVENT_
 
+  close:
```
Comment (kasperl) [inline lib/ble.toit:776]:
```suggestion
  close -> none:
```

File: lib/ble.toit [hunk 12]
```diff
@@ -721,11 +858,29 @@ class Peripheral extends Resource_:
...
+
   constructor .adapter bonding/bool secure-connections/bool:
     resource := ble-create-peripheral-manager_ adapter.resource-group_ bonding secure-connections
     super adapter.resource-group_ resource --auto-release
     resource-state_.wait-for-state STARTED-EVENT_
 
+  /**
+  Closes the peripheral manager and all its services.
+  */
+  close:
```
Comment (kasperl) [inline lib/ble.toit:871]:
```suggestion
  close -> none:
```

File: lib/ble.toit [hunk 13]
```diff
@@ -721,11 +858,29 @@ class Peripheral extends Resource_:
...
+    stop-advertise
+    services := services_
+    services_ = []
+    services.do: | service/LocalService | service.close
+    close_
+
+  /**
+  Removes the given local-service from the list of deployed services.
```
Comment (kasperl) [inline lib/ble.toit:879]:
```suggestion
  Removes the given $local-service from the list of deployed services.
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/ble.toit (hunk 4 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 5 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 6 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 8 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 3 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 2 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 7 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 9 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 10 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 11 above)
Comment (floitsch) [reply lib/ble.toit:792]: done.

File: lib/ble.toit (hunk 12 above)
Comment (floitsch) [reply lib/ble.toit:887]: done.

File: lib/ble.toit (hunk 13 above)
Comment (floitsch) [reply lib/ble.toit]: done.

File: lib/ble.toit (hunk 1 above)
Comment (floitsch) [reply lib/ble.toit]: done.

**Pull request merged**